include/ensmallen_bits/de/de_impl.hpp

```diff
@@ -0,0 +1,129 @@
+
+//!Optimize the function
+template<typename DecomposableFunctionType>
+inline double DE::Optimize(DecomposableFunctionType& function, arma::mat& iterate)
+{
+  // Population Size must be atleast 3 for DE to work
+  if (populationSize < 3)
+  {
+    throw std::logic_error("CNE::Optimize(): population size should be at least"
+        " 3!");
+  }
+
+  // Initialize helper variables
+  fitnessValues.set_size(populationSize);
+  double lastBestFitness = DBL_MAX;
+  arma::mat bestElement;
+
+  // Generate a population based on a Gaussian distribution around the given starting point.
+  // Also finds the best element of the population
+  population = arma::randn(iterate.n_rows, iterate.n_cols, populationSize);
+  for (size_t i = 0; i < populationSize; i++)
+  {
+    population.slice(i) = population.slice(i) + iterate;
+    fitnessValues[i] = function.Evaluate(population.slice(i));
+    if(fitnessValues[i] < lastBestFitness)
+    {
+      lastBestFitness = fitnessValues[i];
+      bestElement = population.slice(i);
+    }
+  }
+
+  Info << "DE initialized successfully. Optimization started."
+  << std::endl;
+
+  // Iterate until maximum number of generations are completed
+  for (size_t gen = 0; gen < maxGenerations; gen++)
+  {
+
+    Info << "Generation: " << gen<< "\nBest Fitness: " << lastBestFitness << std::endl;
+
+    // Generate new populationbased on /best/1/bin strategy
+    for (size_t member = 0; member < populationSize; member++)
+    {
+	    iterate = population.slice(member);
+
+      // Generate two different random numbers to choose two random members
+	    int l=0, m=0;
+	    do
+	    {
+	      l = arma::as_scalar(arma::randi<arma::uvec>(1, arma::distr_param(0, populationSize - 1)));
+	    }
+	    while(l == member);
+	    do
+	    {
+	      m = arma::as_scalar(arma::randi<arma::uvec>(1, arma::distr_param(0, populationSize - 1)));
+	    }
+	    while(m == member && m == l);
+
+	    // Generate new "mutant" from two randomly chosen members
+	    arma::mat a = population.slice(l);
+	    arma::mat b = population.slice(m);
+	    arma::mat mutant = bestElement + differentialWeight * (a - b);
+
+	    // Perform crossover
+	    arma::mat cr = arma::randu(iterate.n_rows);
+	    for (size_t it = 0; it<iterate.n_rows; it++)
+	    {
+	      if (cr[it] >= crossoverRate)
+	      {
+	        mutant[it] = iterate[it];
+	      }
+	    }
+
+	    double iterateValue = function.Evaluate(iterate);
+	    double mutantValue = function.Evaluate(mutant);
+
+	    // Replace the current member if mutant is better
+	    if (mutantValue < iterateValue)
+	    {
+	      iterate = mutant;
+	      iterateValue = mutantValue;
+	    }
+
+	    fitnessValues[member] = iterateValue;
+
+	    population.slice(member) = iterate;			
+    }
+
+    // Update helper variables 
+    lastBestFitness = fitnessValues.min();
+    for (size_t it = 0; it < populationSize; it++)
+      if (fitnessValues[it] == lastBestFitness)
+        bestElement = population.slice(it);
```

Comment:
Could we use `index_min()` here?

Reply:
arma::vec doesn't have a index_min() member function, and if i try to use the standalone function it gives me the error "index_min() is not a member of arma". That's why i did it this way.

Reply:
You should be able to fully specify the name with `arma::index_min()`. :)

Reply:
Yes, doing that is giving me that error.
The offending line of code:
`arma::uword index = arma::index_min(fitnessValues);`
The error message:
error: ‘index_min’ is not a member of ‘arma’

Reply:
Oh, sorry, this is because the minimum supported version of Armadillo doesn't have that.  Sorry about that.  In this case the loop you have is fine but you might consider adding a `break` inside the `if` so it doesn't perform unnecessary iterations. :+1: